src/Query.ts

```diff
@@ -0,0 +1,164 @@
+import { TaskEither, taskEither } from 'fp-ts/lib/TaskEither';
+import { Cache } from './Cache';
+import { mapWithKey, sequence, map as mapRecord } from 'fp-ts/lib/Record';
+import {
+  Strategy,
+  JSON,
+  setoidStrict,
+  setoidShallow,
+  setoidJSON,
+  refetch
+} from './Strategy';
+import { Setoid } from 'fp-ts/lib/Setoid';
+import { CacheValue, getSetoid } from './CacheValue';
+import {
+  EnforceNonEmptyRecord,
+  ObservableQueries,
+  ProductA,
+  ProductL,
+  ProductP
+} from './util';
+
+export type Fetch<A, L, P> = (input: A) => TaskEither<L, P>;
+
+interface BaseQuery<A, L, P> {
+  _A: A;
+  _L: L;
+  _P: P;
+  run: (input: A) => TaskEither<L, P>;
+  invalidate: (input: A) => TaskEither<L, P>;
+  gc: () => void;
+}
+
+function queryPhantoms<A, L, P>(): { _A: A; _L: L; _P: P } {
+  return null as any;
+}
+
+export interface CachedQuery<A, L, P> extends BaseQuery<A, L, P> {
+  type: 'cached';
+  cache: Cache<A, L, P>;
+}
+
+export interface Composition<A, L, P> extends BaseQuery<A, L, P> {
+  type: 'composition';
+  master: ObservableQuery<A, L, unknown>;
+  slave: ObservableQuery<unknown, L, P>;
+}
+
+export interface Product<A, L, P> extends BaseQuery<A, L, P> {
+  type: 'product';
+  queries: Record<string, ObservableQuery<A[keyof A], L, P[keyof P]>>;
+}
+
+export type ObservableQuery<A, L, P> =
```

Comment:
what is making the query `observable`? (sorry never worked with them so I'm missing some steps)

Reply:
the fact that you can call `observable` on it and obtain an `Observable<QueryResult<A, L, P>>` back.
The "observability core" is part of the `ObservableCache.ts` file: cached queries ("leaf") have an observable cache attached, compositions and products just merge together events coming from cached queries end re-expose them composed